# Lista de ejercicios 2

1. Supongamos que `x` es un vector numérico. **Explica en detalle**, como las siguientes expresiones son evaluadas y que valores toman:


  * sum(!is.na(x))

In [1]:
sum(!is.na(x))
# !is.na(x) devuelve un vector lógico de TRUE y FALSE los cuáles son coercionados a 
# los valores de 1 y 0 respectivamente al afectarlos la función sum().

[1] 4

 * c(x,x[-(1:length(x))])

In [2]:
x <- c(2, 3, 5, 7)
c(x, x[-(1: length(x))])
# el vector x al encerrar a índices con signo negativo desde 1 hasta su longitud 
# devuelve el vector numeric(0) el cual al enlazarlo con x mediante la función c()
# nos devuelve el mismo vector x.


[1] 2 3 5 7

 * x[length(x) + 1]/length(x)

In [3]:
x[length(x) + 1]/length(x)
# Al insertar como índice de x a un valor mayor que su longitud, el valor retornado
# es NA, el cual al ser dividido entre el valor de su longitud resulta nuevamente NA

[1] NA

 * sum(x > mean(x))

In [4]:
sum(x > mean(x))
# Esta línea de código nos devuelve la suma de los valores de x que resultan ser 
# mayores a su media.

[1] 2

2.La función seno hiperbólico es definida como:
$$sinh x = \frac{e^x - e^{-x}}{2}$$

donde $e^x$ es la función exponencial. Usando solo operaciones aritméticas y la función `exp`, escribe una función `vectorizada` en R, que calcule el $\sinh$.

In [5]:
# Definimos la función sen_hiper()
sin_hyper <- function(x){(exp(x) + exp(-x))/2}
# Esta funcióncalcula el valor del seno hiperbólico para cada componente 
# del vector x

3.Crea un vector de los valores de $e^x\cos(x)$ en $x = 3,3.1, 3.2, \dots, 6$

In [6]:
a <- seq(from = 3, to = 6, by = 0.1) 
# almacenamos el vector que va desde 3 hasta 6 avanzando de 0.1 en 0.1 en a.
exp(a)*cos(a)
#Gracias a la vectorización de R, la última línea de código nos devuelve el 
# valor de la exponencial de cada componente de a multiplicada por el respectivo
# coseno de componente de a

[1] -19.884531 -22.178753 -24.490697 -26.773182 -28.969238 -31.011186
 [7] -32.819775 -34.303360 -35.357194 -35.862834 -35.687732 -34.685042
[13] -32.693695 -29.538816 -25.032529 -18.975233 -11.157417  -1.362099
[19]  10.632038  25.046705  42.099201  61.996630  84.929067 111.061586
[25] 140.525075 173.405776 209.733494 249.468441 292.486707 338.564378
[31] 387.360340

4.Usa la función `paste` para crear el siguiente vector de caracteres de longitud $30$:
 * $("vector 1", "vector 2", \dots, "vector 30")$. Se notar el espacio entre `vector` y el número.

In [7]:
(paste("vector", 1:30))
# Concatenamos los vectores con la función paste(), después de
# coercionarlos a caracteres, dicha función preserva el espacio entre 
# vectores

[1] "vector 1"  "vector 2"  "vector 3"  "vector 4"  "vector 5"  "vector 6" 
 [7] "vector 7"  "vector 8"  "vector 9"  "vector 10" "vector 11" "vector 12"
[13] "vector 13" "vector 14" "vector 15" "vector 16" "vector 17" "vector 18"
[19] "vector 19" "vector 20" "vector 21" "vector 22" "vector 23" "vector 24"
[25] "vector 25" "vector 26" "vector 27" "vector 28" "vector 29" "vector 30"

* $("fn1", "fn2", \dots, "fn30")$. En este caso no hay espacio entre `fn` y el número.


In [8]:
paste0("fn", 1:30)
# paste0() funciona de manera idéntica a paste() solo que no preserva el espacio
# entre las cadenas concatenadas
# en ambas líneas de código se ha hecho uso de recycling

[1] "fn1"  "fn2"  "fn3"  "fn4"  "fn5"  "fn6"  "fn7"  "fn8"  "fn9"  "fn10"
[11] "fn11" "fn12" "fn13" "fn14" "fn15" "fn16" "fn17" "fn18" "fn19" "fn20"
[21] "fn21" "fn22" "fn23" "fn24" "fn25" "fn26" "fn27" "fn28" "fn29" "fn30"

5.¿Qué retorna $\texttt{DF <- D(expression(cos(x)/sin(x)), "x")}$. Si ejecutamos el comando 
$\texttt{x <- pi/4}$, que retorna $\texttt{eval(DF)}$?.


In [9]:
DF <- D(expression(cos(x)/sin(x)), "x")
x <- pi/4
eval(DF)

[1] -2

6.¿Puedes explicar los siguientes resultados?

 * `x <- c(0,7,8)`
 
  Se almacenan los valores `0, 7, 8` en las posiciones `0, 1, 2` de `x`
  
 * `x[0.9999999999999999]` 16 veces nueve
 * `numeric(0)`
 
  R solo puede almacenar a lo más `15` o `16` dígitos, por lo que al ejecutar la línea de código 
 superior, R trucna dicho número a `0`, pues es menor a `1`, es decir, `R` lee `x[0]` por lo que 
 la salida es `numeric(0)` una cadena vacía.
 

 * `x[0.99999999999999999]` 17 veces nueve
 
 * `0`
 
 Aquí, `R` se encuentra con `17` nueves, por lo que `R` los mira como si fueran `1`, entonces `R` lee `x[1]`

7.Usando $\texttt{rep()}$ y $\texttt{seq()}$, crea los siguientes vectores:


`0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3 4 4 4 4 4`

y

`1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5`


In [10]:
rep(seq(0, 4), each = 5)
rep(seq(5), 5)

[1] 0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3 4 4 4 4 4

[1] 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5

8.Usando la función $\texttt{cumprod}$ o otra relacionada, calcula

$$
1 + \frac{2}{3}+ \biggl(\frac{2}{3}\frac{4}{5}  \biggr) +  \biggl(\frac{2}{3}\frac{4}{5}\frac{6}{7} \biggr) + \cdots + \biggl(\frac{2}{3}\frac{4}{5}\cdots \frac{38}{39} \biggr)
$$

In [11]:
1 + sum(cumprod(seq(2, 38, 2))/cumprod(seq(3, 39, 2)))
# cumprod(seq(2, 38, 2)) nos da los productos acumulativos como 2, 2*4, 2*4*6, ...; 
# mientras que cumprod(seq(3, 39, 2) nos retorna 3, 3*5, 3*5*7, ...; estas expresiomes
# al ser divididas son generadas los sumandos de la expresión pedida, por lo que para 
# sumarlos llamamos a la función sum() y le añadimos 1.

[1] 6.976346

9.En primer lugar, lanzamos una moneda con dos posibles resultados: o bien cae cara o sello. A continuación, tiramos un dado con $6$ posibles resultados: $1,2,3,4,5$ y $6$. Genera todos los resultados posibles de esas acciones.


In [12]:
expand.grid(S1 = c("C", "S"), S2 = c(1, 2, 3, 4, 5, 6))

S1 S2
1  C  1 
2  S  1 
3  C  2 
4  S  2 
5  C  3 
6  S  3 
7  C  4 
8  S  4 
9  C  5 
10 S  5 
11 C  6 
12 S  6

10.Tienes un conjunto de colores para elegir:

Ahora debes  elegir $3$ colores y no puedes "elegir el mismo color más de una vez. Enumera todas las combinaciones posibles.

In [13]:
combn(c("rojo", "azul", "verde", "blanco", "negro", "amarillo"), 3)

[,1]  [,2]   [,3]  [,4]     [,5]   [,6]  [,7]     [,8]   [,9]     [,10]   
[1,] rojo  rojo   rojo  rojo     rojo   rojo  rojo     rojo   rojo     rojo    
[2,] azul  azul   azul  azul     verde  verde verde    blanco blanco   negro   
[3,] verde blanco negro amarillo blanco negro amarillo negro  amarillo amarillo
     [,11]  [,12] [,13]    [,14]  [,15]    [,16]    [,17]  [,18]    [,19]   
[1,] azul   azul  azul     azul   azul     azul     verde  verde    verde   
[2,] verde  verde verde    blanco blanco   negro    blanco blanco   negro   
[3,] blanco negro amarillo negro  amarillo amarillo negro  amarillo amarillo
     [,20]   
[1,] blanco  
[2,] negro   
[3,] amarillo

11.Tienes el mismo conjunto de colores y puedes elegir $1$, $2$ o $3$. Haz una lista de todas las opciones posibles.


In [14]:
# Si eligo un color:
combn(c("rojo", "azul", "verde", "blanco", "negro", "amarillo"), 1)
# Si eligo dos colores:
combn(c("rojo", "azul", "verde", "blanco", "negro", "amarillo"), 2)
# Si eligo tres colores:
combn(c("rojo", "azul", "verde", "blanco", "negro", "amarillo"), 3)

[,1] [,2] [,3]  [,4]   [,5]  [,6]    
[1,] rojo azul verde blanco negro amarillo

[,1] [,2]  [,3]   [,4]  [,5]     [,6]  [,7]   [,8]  [,9]     [,10]  [,11]
[1,] rojo rojo  rojo   rojo  rojo     azul  azul   azul  azul     verde  verde
[2,] azul verde blanco negro amarillo verde blanco negro amarillo blanco negro
     [,12]    [,13]  [,14]    [,15]   
[1,] verde    blanco blanco   negro   
[2,] amarillo negro  amarillo amarillo

[,1]  [,2]   [,3]  [,4]     [,5]   [,6]  [,7]     [,8]   [,9]     [,10]   
[1,] rojo  rojo   rojo  rojo     rojo   rojo  rojo     rojo   rojo     rojo    
[2,] azul  azul   azul  azul     verde  verde verde    blanco blanco   negro   
[3,] verde blanco negro amarillo blanco negro amarillo negro  amarillo amarillo
     [,11]  [,12] [,13]    [,14]  [,15]    [,16]    [,17]  [,18]    [,19]   
[1,] azul   azul  azul     azul   azul     azul     verde  verde    verde   
[2,] verde  verde verde    blanco blanco   negro    blanco blanco   negro   
[3,] blanco negro amarillo negro  amarillo amarillo negro  amarillo amarillo
     [,20]   
[1,] blanco  
[2,] negro   
[3,] amarillo

12.
(a) Sea $X$ el número de $\texttt{"unos"}$ obtenidos en `12` lanzamientos  de un dado. Entonces $X$ tiene una  distribución binomial $(n=12, p=1/3)$ . Calcule una tabla de probabilidades binomiales para $x=0, 1, \dots ,12$ por dos métodos:
 * Usando la fórmula para la densidad:
$$P(X = K)=\binom{n}{k}p^{k}(1 -p)^{n -k}$$
y la aritmética en R. Usa $\texttt{0:12}$ para la secuencia de $x$ valores y la función $\texttt{choose}$ para calcular los coeficientes binomiales $\binom{n}{k}$.

 * Usando la función $\texttt{dbinom}$ de **R** y comparar tus resultados con ambos métodos.

In [15]:
# Definimos las siguientes variables
n = 12
p = 1/3
k = 0:12
# Calculamos la densidad de nuestra variable binomial usando choose()
choose(n,k)*(p**k)*((1 - p)**(n - k))
# Ahora usamos la función dbinom()
dbinom(k, n, p)
# Comparamos ambos resultados, primero solo viendo si son igualment cercanos con 
# all.equal(), y nos retorna TRUE pues los vvalores truncados son los mismos.
all.equal(dbinom(k, n, p), choose(n,k)*(p**k)*((1 - p)**(n - k)))

# Sin embargo, identical() nos retorna FALSE ya que los valores difieren al final 
# sus cifras decimales.
identical(dbinom(k, n, p), choose(n,k)*(p**k)*((1 - p)**(n - k)))


[1] 7.707347e-03 4.624408e-02 1.271712e-01 2.119520e-01 2.384460e-01
 [6] 1.907568e-01 1.112748e-01 4.768921e-02 1.490288e-02 3.311751e-03
[11] 4.967626e-04 4.516023e-05 1.881676e-06

[1] 7.707347e-03 4.624408e-02 1.271712e-01 2.119520e-01 2.384460e-01
 [6] 1.907568e-01 1.112748e-01 4.768921e-02 1.490288e-02 3.311751e-03
[11] 4.967626e-04 4.516023e-05 1.881676e-06

[1] TRUE

[1] FALSE

(b) Sea $X$ el número de $\texttt{"unos"}$ obtenidos en `12` lanzamientos  de un dado. Entonces $X$ tiene una  distribución binomial $(n=12, p=1/3)$. Calcule el CDF para $x=0, 1, \dots ,12$ por dos métodos:
 * Usando la función $\texttt{cumsum}$ y el resultado del ejercicio anterior.
 * Con el uso de la función $\texttt{pbinom}$. ¿Qué es $P(X >7)$?.

In [16]:
# Independientemente de si usamos choose(n,k)*(p**k)*((1 - p)**(n - k)) o 
# dbinom(k, n, p) el último termino nos da 1 como debería ser en el cálculo 
# de la CDF.
cumsum(dbinom(k, n, p))
cumsum(choose(n,k)*(p**k)*((1 - p)**(n - k)))

# Igualmente con la función pbinom() 
# para el calculo del CDF con lo uqe el último termino resulta ser 1, como debería ser
pbinom(k, n, p)

# P(X > 7) nos brinda la probabilidad de uqe nuestra variable aleatoria sea mayor 
# a 7, y su valor esta dado por:
sum(dbinom(k, n, p)[9:13])

[1] 0.007707347 0.053951426 0.181122646 0.393074678 0.631520714 0.822277544
 [7] 0.933552360 0.981241568 0.996144445 0.999456196 0.999952958 0.999998118
[13] 1.000000000

[1] 0.007707347 0.053951426 0.181122646 0.393074678 0.631520714 0.822277544
 [7] 0.933552360 0.981241568 0.996144445 0.999456196 0.999952958 0.999998118
[13] 1.000000000

[1] 0.007707347 0.053951426 0.181122646 0.393074678 0.631520714 0.822277544
 [7] 0.933552360 0.981241568 0.996144445 0.999456196 0.999952958 0.999998118
[13] 1.000000000

[1] 0.01875843

13.
 * ¿Cuál es la probabilidad de que una persona sea menor o igual a `1.90` m? Usa $\texttt{pnorm}$.

In [17]:
pnorm(1.90, 0, 1)

[1] 0.9712834

* ¿Cuál es la probabilidad de que una persona sea más alta o igual a `1.60` m?.  Usa $\texttt{pnorm}$.

In [18]:
pnorm(160, 0, 100, lower.tail = FALSE) + dnorm(160, 0, 100)
# El último arguemento de pnorm hace que se consideren los 
# valores mayores que 160

[1] 0.0559085

14.¿Cuál es la probabilidad de que una persona espere menos de $10$ minutos?. Utiliza $\texttt{pexp}$.


In [19]:
# Nuestras unidadeds referenciales son los minutos, pexp() asumirá por defecto
# que rate = 1
pexp(10)

[1] 0.9999546

15.Ejecuta las siguientes líneas que crean dos vectores de enteros aleatorios que se eligen con reemplazo de los enteros $0, 1,\dots, 999$ . Ambos vectores tienen una longitud de $250$.


In [20]:
(set.seed(50))
(xVec <- sample(0:999, 250, replace=T))
(yVec <- sample(0:999, 250, replace=T))

# set.seed(50) hace que cada muestra aleatoria sea la misma, sin cambiar cada vez que los vectores xVec y yVect 
# llamadas, pues de hecho deberían cambiar sus valores debido a la aleatoriedad que genra sample().

NULL

[1] 708 437 200 767 513  44 699 646  42 107 390 269 640  77 277 676 835 364
 [19]  74 168 616 193 710 842 309 650 577 257 324 368 358 408 437 618 222 627
 [37] 121 701 373 458 363 836 278  93  55 700 954 458 713 803 996 765 639 299
 [55] 358 425 715 525 511 266 578 655 197 585 129  38 724  61 136 944 507 995
 [73] 661  74 967 148 657 956 652 956 543  17 339 469 544  19   1 680 537 645
 [91] 691 688 828 760  48 294  69 807 311 668 505 964 632   8  24 862  10 614
[109] 840 353 878  72 193 113  82 322  91 789 444 986 624  18 537 554 515 460
[127] 263  42  76 256 359 189 807 457  99 274 543 324 176 477 541 160 260 174
[145]  48 415 707 625 530 407 216 224 395 977 828 461 148 293 660  38 137 224
[163] 852 743 683 545 353 371 866 452 811 768 339 203 478  49  20 880 480 996
[181] 894 357 900 603 667 787 972 457 467 324 928 109 365 987 572 280 113 702
[199] 963 405  63 621 517 446 533 190 638 275 865 435 501 669 124  14 920 308
[217]  84 523   5 863 860 120 206 399  29 256 678  59 497 188 127 258 376 171
[235] 781 870 110 957 285 382  34 403 631 197 179 545 123 760 238 178

[1] 709 871 315 517 621 930 437 948 157 783 878 471 671  91 415 860 273 768
 [19] 581 381  47 347 229   4 279 411 698 974 554 279 216 855 813 776 218 721
 [37] 538 332  31 460 532 917 985  95 705 248 247 884 317 840  94 288  43 575
 [55] 687 174 213 957 955 786 938 428 930 101 330 641 615 988 500 285  28 881
 [73] 106 329 398 414 542 570 881 997 221 488 117 299 484 823 428 791 133  50
 [91] 246  72 520 643 779 693 845 296 441 553 815 752 465  18 766  87 635 257
[109] 993 368 919 116 224 686 473 151 512 635 613 660 310 419 800 428 743 282
[127] 965  44 330  19 743 615 489 615 194 803 948 760 604 193 409 800 772 133
[145] 175 593 184 516 287 863 902 195 220 689 309  14 881 941 924 593 693 280
[163] 835 632 225 398 872 876 358 187 211 850 961 681 791 947 117 915 222 712
[181] 665 921 798 167 421 268 866 503 828 942 589 521 320 424  13 482 498 509
[199] 216   0  78 488 841 645 681 827  83 273 421 277 884  67 890 970 400  10
[217] 469 290 632 717 529 426 127 846  49 952 609  99 284 824 598 695  63 293
[235] 325 295 675 777 813 557 792 580 783  72 611 853 738 345 668 791